### PR TITLE
[GPII-2967]: Add gcp-stackdriver-alerting module, generic policies

### PR DESCRIPTION
**Pull request by natarajaya**
Here is proposed solution that uses `google-cloud-monitoring` gem to manage Stackdriver resources.
It supports all basic functionality to create / update / delete notification channels, uptime checks and alerting policies from json configs.

To test basic functionality I added only uptime checks for preferences and flow manager, policies based on them and single notification channel – my email.

Current issues:
* A lot of ruby code (but [Terraform's `monitoring_alert_policy` object](https://github.com/modular-magician/terraform-provider-google/blob/927501f0be9e8d5a4f969d727ee346c25117d72e/google/resource_monitoring_alert_policy.go) got 10x more code)! I tried to keep it as simple as possible.
* Only email addresses are supported as notification channels. This can be fixed relatively easy.
* All notification channels are getting applied to every policy. No way to specify notification channels per alert policy, no way to specify notification channels per environment.
* The process to create resource primitives is pretty difficult. There is no way to export monitoring resources directly from Stackdriver Dashboard, there are no corresponding `gsutil` commands as well. The only way to manually configure, let's say, alert policy in Dashboard and then get json config for it is via API. I used the client script from the module, to dump resource objects, but there is still some manual editing involved.

### CODE REVIEW

**Comment by natarajaya** [comment]
@amatas @stepanstipl @mrtyler This is ready for review now. I updated description with short progress summary.

**Comment by stepanstipl** [comment]
I have a question - I understand Terraform does not support all the features we require yet, but even using just the alert policy (`google_monitoring_alert_policy`) would seem like a step towards less code on our side. Is there a reason why not to use it?

**Comment by natarajaya** [comment]
@stepanstipl My main motivation was consistency: in case we used Terraform's alert policy implementation, we would have had some of the primitives described with their json configs and some with Terraform resources. We would also need a way to pass notification channels ids (can not be stored in json configs) from ruby code to Terraform google_monitoring_alert_policy resources. But I don't have strong opinion about this.

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  stackdriver_notification_channels = {}
+  processed_notification_channels = {}
+
+  notification_channel_service_client.list_notification_channels(formatted_parent).each do |notification_channel|
+    puts notification_channel.to_json if @debug_mode
+    stackdriver_notification_channels[get_notification_channel_attribute(notification_channel)] = notification_channel
+  end
+
+  notification_channels.each do |notification_channel|
+    notification_channel_attribute = get_notification_channel_attribute(notification_channel)
```
Comment (mrtyler) [inline gcp/modules/gcp-stackdriver-alerting/client.rb]: I think I understand what you're trying to do with `attribute` here, but it's such an abstract name that I found it hard to follow the logic. How about replacing `attribute` with `name` or maybe `identifier`?

**Comment by mrtyler** [comment]
Second, a strategic pass.

1. This is a moderate amount of custom Ruby code[*], including a simple implementation of Terraform-style idempotence and destruction. Should we invest the time and effort now to add unit tests?

2. I haven't used Stackdriver monitoring very much yet, but so far I've found it a little underwhelming.  We need a moderate amount of custom Ruby code to manage its monitoring/alerting resources. Sergey couldn't find any meaningful example alerts to copy. We (maybe just I?) suspect that Stackdriver will be less flexible than Prometheus+Grafana. Should we consider switching back to Prometheus+Grafana for monitoring/alerting?
Some upsides: easier to install (Helm chart or maybe that Google-managed one that Sergey found), can use alerts / dashboards from the community, possibly more flexible. The biggest downside: self-hosted vs Google-worries-about-it. Is maintaining and operating a Prometheus+Grafana cluster more effort than maintaining this custom Ruby code?



[*] I feel like I say this a lot even though I was in the conversations where we discussed why we need  this custom Ruby code :) -- in this case, lack of Terraform support for Stackdriver alerting resources. Still, for the record: We want to minimize custom code because code is a liability. Also, the more custom code we have, the "stranger" our project is and the less we benefit from the broader ecosystem.

---
* custom ruby code, implementing idempotency
** with no unit tests
* no canned examples to work from

prometheus and friends instead?

**Comment by stepanstipl** [comment]
> My main motivation was consistency: in case we used Terraform's alert policy implementation, we would have had some of the primitives described with their json configs and some with Terraform resources. We would also needed to find a way to pass notification channels ids (can not be stored in json configs) from ruby code to Terraform google_monitoring_alert_policy resources. But I don't have strong opinion about this.

@natarajaya I agree with the concern for consistency, but as the long-term approach I believe is to use Terraform, I would be inclined to use the `google_monitoring_alert_policy` and other resources as they will becoming available. I think this would also be more along the lines with @mrtyler's points on limiting custom code.

As to @mrtyler's 2) - my feeling is that we discussed this greatly and I would stay with Stackdriver, unless we have some specific things that can't be done. It seems to me that Stackdriver provides enough in terms of features with the benefit of managed service. Yes, it's true that we won't get completely ready setup out of the box, but I believe that vast majority, if not all, of the alerts are transferable (even if not in a direct copy way). If not - I would like to hear specific examples. The amount of custom code will decrease as better support is coming to TF.

Btw. there are other cool features that Stackdriver can do - Error reporting and Distributed tracing for example, upcoming integration with Istio and Service Monitoring...

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+def process_notification_channels(notification_channels = [])
+  notification_channel_service_client = Google::Cloud::Monitoring::NotificationChannel.new(version: :v3)
+  formatted_parent = Google::Cloud::Monitoring::V3::NotificationChannelServiceClient.project_path(@project_id)
+
+  stackdriver_notification_channels = {}
+  processed_notification_channels = {}
+
+  notification_channel_service_client.list_notification_channels(formatted_parent).each do |notification_channel|
+    puts notification_channel.to_json if @debug_mode
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: Sounds good!

File: gcp/modules/gcp-stackdriver-alerting/main.tf [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    project_id  = "${var.project_id}"
+    domain_name = "${var.domain_name}"
+  }
+}
+
+resource "null_resource" "apply_stackdriver_alerting" {
+  depends_on = ["template_dir.resources"]
+
+  triggers = {
+    nonce = "${var.nonce}"
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/main.tf]: It is possible, but we would need to save rendered templates in `destination_dir` locally (using another docker volume probably), otherwise resource re-creation will be triggered on every run. I personally think it does not worth the effort.

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+      notification_channel = notification_channel_service_client.create_notification_channel(formatted_parent, notification_channel)
+    end
+
+    processed_notification_channels[notification_channel_attribute] = notification_channel["name"]
+  end
+
+  stackdriver_notification_channels.each do |name, notification_channel|
+    notification_channel_attribute = get_notification_channel_attribute(notification_channel)
+
+    unless processed_notification_channels.include? notification_channel_attribute or @debug_mode
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: It is the other way around – enabling debug mode disables deletion of Stackdriver resources.

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    unless processed_alert_policies.include? alert_policy_attribute or @debug_mode
+      puts "Deleting alert policy \"#{alert_policy_attribute}\"..."
+      alert_policy_service_client.delete_alert_policy(alert_policy["name"])
+    end
+  end
+
+  return processed_alert_policies
+end
+
+unless @destroy_triggered
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: There is no `main` method in Ruby. I'll keep only definitions in `client.rb` and move in-line code into `local-exec` resources.

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 6]
```diff
@@ -0,0 +1,172 @@
+require "json"
+require "google/cloud/monitoring"
+
+@project_id = ENV['PROJECT_ID']
+@destroy_triggered = ENV['DESTROY']
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: Makes sense. I don't think we need `DESTROY` var anymore, I'll rename `DEBUG` into `STACKDRIVER_DEBUG`.

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+@debug_mode = ENV['DEBUG']
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: Same here.

File: gcp/modules/gcp-stackdriver-alerting/client.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: Replied in https://github.com/gpii-ops/gpii-infra/pull/183#discussion_r227588841.

File: gcp/modules/gcp-stackdriver-alerting/resources/alert_policies/container_resource_limit_utilization.json [hunk 8]
```diff
@@ -0,0 +1,59 @@
+{
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/resources/alert_policies/container_resource_limit_utilization.json]: Would you like it to be a part of `gcp/README.md` or should I add a separate readme file in `gcp/modules/gcp-stackdriver-alerting`?

File: gcp/modules/gcp-stackdriver-alerting/client.rb (hunk 1 above)
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: These functions are needed because we can't rely on Stackdriver resource names and want to be able to identify resources by some other attribute / set of attributes. But yeah, I don't like `_attribute` postfix either. `_identifier` probably makes more sense.

**Comment by natarajaya** [comment]
@mrtyler @stepanstipl Thanks for your reviews!

> Should we invest the time and effort now to add unit tests?

IMO, unit tests are most efficient in test-driven development. Now we'll have to not only write tests, but also adjust code to work with the tests. It may be not so easy, considering that all the code does is calling Google Cloud API via `google-cloud-monitoring` gem. We can add tests to validate resource configs, though I am not sure if it worth the effort.

> I haven't used Stackdriver monitoring very much yet, but so far I've found it a little underwhelming

Well, while I agree that in general we'll have more flexibility with Prometheus + Grafana (which comes with a cost of maintaining and running them), Stepan raises a good point – is there a functionality we're looking for and that can't be achieved with Stackdriver? If the answer on that question is no, I think we should stick with Stackdriver, specially knowing that we can ditch Ruby client entirely, once Terraform implements all resource primitives that we need.

> I agree with the concern for consistency, but as the long-term approach I believe is to use Terraform, I would be inclined to use the google_monitoring_alert_policy and other resources as they will becoming available

I tried to implement the option with `google_monitoring_alert_policy` objects managed in Terraform and it looked pretty messy. The only reasonable way to get processed notification channel identifiers from Ruby client that I found, is by using external data source. Which creates issues with dependencies, because resources needs to be created and deleted in opposite order (you can't apply alert policies before notification channels, and you can't delete notification channels while there are policies that relies on them). I think we should be fine managing all Stackdriver resources in Ruby for now.

**Comment by natarajaya** [comment]
This is getting close to be ready for merging.

Pending issues:
* Instructions on how to create / debug json resource configs needs to be added to `gcp/README.md` or separate readme file in `gcp/modules/gcp-stackdriver-alerting`.
* There is single notification channel: email. Is this sufficient for the beginning? I used the address `alerts+<%= ENV["TF_VAR_environment"] %>@raisingthefloor.org` from AWS part, but looks like something needs to be done on [Google Groups side of things](https://groups.google.com/a/raisingthefloor.org/forum/#!pendingmsg/alerts).

File: gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
 {
   "type": "email",
   "labels": {
-    "email_address": "sergey@raisingthefloor.org"
+    "email_address": "alerts+${env}@raisingthefloor.org"
```
Comment (mrtyler) [inline gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json]: This is fine, but I think it might be helpful to distinguish between different dev environments (e.g. Alice may want to subscribe to alerts@ and filter out everything but alerts+dev-alice; we all want to filter out alerts+dev-bob because Bob is always breaking stuff in his environment).

How about `var.project_id` or `var.domain_name` instead?

File: gcp/modules/gcp-stackdriver-alerting/main.tf [hunk 10]
```diff
@@ -28,8 +30,14 @@ resource "null_resource" "apply_stackdriver_alerting" {
     command = <<EOF
       export PROJECT_ID=${var.project_id}
       export GOOGLE_CLOUD_KEYFILE=${var.serviceaccount_key}
-      # export DEBUG=1
-      ruby ${path.module}/client.rb
+      # export STACKDRIVER_DEBUG=1
+      ruby -e '
+        require "${path.module}/client.rb"
+        resources = read_resources
```
Comment (mrtyler) [inline gcp/modules/gcp-stackdriver-alerting/main.tf]: I like this refactor, but I don't like this logic being in the Terraform file. How about?

```
require client.rb
apply_resources
```

and

```
require client.rb
destroy_resources
```

This keeps the ruby logic in the ruby code while clarifying and simplifying everything by removing the `DESTROY=1` trick.

File: gcp/modules/gcp-stackdriver-alerting/client.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (mrtyler) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: I still find this behavior pretty unintuitive. (I also find `unless X or Y` hard to reason about but maybe that's just me.)

If skipping deletion in debug mode is desirable, I might do it like this:
```
unless X
  if @debug_mode
    puts "[DEBUG] Skipping deletion of #{name} because @debug_mode is set"
  else
    do_the_thing
  end
end
```

Not a big deal, ok to skip if you think this is unlikely to bite anyone.

File: gcp/modules/gcp-stackdriver-alerting/resources/alert_policies/container_resource_limit_utilization.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (mrtyler) [reply gcp/modules/gcp-stackdriver-alerting/resources/alert_policies/container_resource_limit_utilization.json]: Putting the documentation in the module seems reasonable. However I worry about discoverability (currently we don't have any documentation at the module level; only at the top level and in shared/charts/).

So maybe the bulk of the documentation goes in a new gcp-stackdriver-alerting/README.md, which is linked under a heading "How do I create/modify Stackdriver Alerts?" from gcp/README.md?

**Comment by mrtyler** [comment]
> There is single notification channel: email. Is this sufficient for the beginning? I used the address alerts+<%= ENV["TF_VAR_environment"] %>@raisingthefloor.org from AWS part,

I think this is all good.

> but looks like something needs to be done on Google Groups side of things.

I think you figured it out (probably that you had to approve a new sender to the list).

Can you document the process somewhere similar to https://github.com/gpii-ops/gpii-infra/blob/master/aws/README.md#alerts-and-notifications?

**Comment by mrtyler** [comment]
Thanks Stepan and Sergey for your comments and the longer discussion about unit testing / TDD in today's retrospective. I'm fine with the consensus on strategy which I would summarize as:
* Testing is good. But we don't think unit testing this ruby code has the best ROI right now, so we won't do it.
* Stick with Stackdriver and this approach to configuring it.

File: gcp/modules/gcp-stackdriver-alerting/client.rb [hunk 11]
```diff
@@ -2,27 +2,42 @@
...
-@destroy_triggered = ENV['DESTROY']
-@debug_mode = ENV['DEBUG']
-@resource_dir="#{File.expand_path(File.dirname(__FILE__))}/resources_rendered"
+@debug_mode = ENV['STACKDRIVER_DEBUG']
 
-def get_notification_channel_attribute(notification_channel)
+def read_resources(resource_dir = "")
+  resource_dir = "#{File.expand_path(File.dirname(__FILE__))}/resources_rendered" if resource_dir.empty?
+
+  Dir.chdir(resource_dir)
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/client.rb]: Block form looks better, thanks!

File: gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json (hunk 9 above)
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json]: I think `alerts+prd.gcp.gpii.net@raisingthefloor.org` is more readable than `alerts+gpii-gcp-prd@raisingthefloor.org`, will update to domain name.

File: gcp/modules/gcp-stackdriver-alerting/main.tf (hunk 10 above)
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/main.tf]: There is no `DESTROY=1` trick anymore, but keeping all logic in `client.rb` file sounds good anyways.

**Comment by natarajaya** [comment]
@mrtyler I addressed issues from all your comments, added "Monitoring and alerting" section into GCP readme.

**Comment by amatas** [comment]
LGTM

**Comment by stepanstipl** [comment]
lgtm

File: gcp/README.md [hunk 12]
```diff
@@ -153,6 +153,26 @@ rake sh["gcloud services disable container.googleapis.com"]
 This happens due to limitation of maximum 10 keys per ServiceAccount.
 If you see this error during any `rake` execution, run `rake destroy_sa_keys` and then try again.
 
+### Monitoring and alerting
+
+We use [Stackdriver Beta Monitoring](https://cloud.google.com/monitoring/kubernetes-engine/) to collect various system metrics, navigate through them with [Kubernetes Dashboard](https://app.google.stackdriver.com/kubernetes), and send alerts when they violate thresholds that are being set by [Stackdriver Alerting Policies](https://app.google.stackdriver.com/policies).
+
+Due to the lack of Terraform integration we use [Ruby client](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-alerting/client.rb) to apply / update / destroy Stackdriver's resource primitives from their [json configs](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-alerting/resources).
+
+**To add new resource / debug existing resources:**
```
Comment (mrtyler) [inline gcp/README.md]: Please make this a subheading `####` for better document organization and to provide an anchor to link to.

File: gcp/README.md [hunk 13]
```diff
@@ -153,6 +153,26 @@ rake sh["gcloud services disable container.googleapis.com"]
...
+1. Enable [debug mode](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-alerting/main.tf#L31) for Stackdriver client.
+1. Add new resource / modify existing resource using corresponding Stackdriver Dashboard. **Supported resources are:**
+   * [Notification channels](https://app.google.stackdriver.com/settings/accounts/notifications/email) (only email notification channel type is currently supported, all notification channels are being applied to every alert policy).
+   * [Uptime checks](https://app.google.stackdriver.com/uptime).
+   * [Alert policies](https://app.google.stackdriver.com/policies).
+1. Run `rake deploy_module['k8s/stackdriver/alerting']`.
+1. You will find json blobs for all existing resource primitives in the output.
+1. To add new resource:
+   * Copy its json blob into proper [resource directory](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-alerting/resources). Give a meaningful name to a new resource file. You can use `jq` to help with formatting.
```
Comment (mrtyler) [inline gcp/README.md]: How do I get a json blob for a newly-created resource?

File: gcp/README.md (hunk 12 above)
Comment (natarajaya) [reply gcp/README.md]: Done.

File: gcp/README.md (hunk 13 above)
Comment (natarajaya) [reply gcp/README.md]: I thought it should be clear from the instructions. Let me rephrase some points a little bit and see if it makes a better explanation.

File: gcp/README.md [hunk 14]
```diff
@@ -159,18 +159,18 @@ We use [Stackdriver Beta Monitoring](https://cloud.google.com/monitoring/kuberne
 
 Due to the lack of Terraform integration we use [Ruby client](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-alerting/client.rb) to apply / update / destroy Stackdriver's resource primitives from their [json configs](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-alerting/resources).
 
-**To add new resource / debug existing resources:**
+#### To add new resource / debug existing resources:
 1. Enable [debug mode](https://github.com/gpii-ops/gpii-infra/blob/master/gcp/modules/gcp-stackdriver-alerting/main.tf#L31) for Stackdriver client.
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply gcp/README.md]: Alright then, agree that solution with `TF_VAR_stackdriver_debug` looks better than solution that involves editing code.

File: gcp/modules/gcp-stackdriver-alerting/main.tf [hunk 15]
```diff
@@ -7,6 +7,10 @@ variable "domain_name" {}
 variable "project_id" {}
 variable "serviceaccount_key" {}
 
+variable "stackdriver_debug" {
```
Comment (mrtyler) [inline gcp/modules/gcp-stackdriver-alerting/main.tf]: Hmm, this variable is never used. I guess we can delete it and rely only on `ENV["TF_VAR_stackdriver_debug"]`? (We need to keep the `TF_VAR_` prefix so that the variable is passed from the user's environment through to processes running in the exekube container.)

File: gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json [hunk 16]
```diff
@@ -0,0 +1,10 @@
+{
+  "type": "email",
+  "labels": {
+    "email_address": "alerts+${domain_name}@raisingthefloor.org"
```
Comment (mrtyler) [inline gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json]: Can you document the process of authorizing new senders somewhere, similar to https://github.com/gpii-ops/gpii-infra/blob/master/aws/README.md#alerts-and-notifications?

**Comment by mrtyler** [comment]
Oh, and

File: gcp/modules/gcp-stackdriver-alerting/main.tf (hunk 15 above)
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/main.tf]: Well, the same way we can get rid of `PROJECT_ID` and `GOOGLE_CLOUD_KEYFILE` vars and just use `TF_VAR_project_id`, `TF_VAR_serviceaccount_key` in the Ruby code. But I don't really like that – it may be hard to keep track of all ENV vars in use by client without them being declared in Terraform module. I'll keep that var and add `export STACKDRIVER_DEBUG=${var.stackdriver_debug}` before Ruby call instead.

File: gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json (hunk 16 above)
Comment (natarajaya) [reply gcp/modules/gcp-stackdriver-alerting/resources/notification_channels/alerts_at_raisingthefloor_org.json]: Done.

**Comment by mrtyler** [comment]
LGTM!
